src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,1138 @@
+{
+   if(mEditController == nullptr)
+      return false;
+
+   using namespace Steinberg;
+   
+   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   {
+      Vst::ParameterInfo parameterInfo;
+      if(mEditController->getParameterInfo(i, parameterInfo) == kResultOk)
+      {
+         if(parameterInfo.flags & Vst::ParameterInfo::kCanAutomate)
+            return true;
+      }
+   }
+
+   return false;
+}
+
+bool VST3Effect::GetAutomationParameters(CommandParameters& parms)
+{
+   if(mEditController == nullptr)
+      return false;
+
+   using namespace Steinberg;
+   
+   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   {
+      Vst::ParameterInfo parameterInfo;
+      if(mEditController->getParameterInfo(i, parameterInfo) == kResultOk)
+      {
+         parms.Write(
+            std::to_string(parameterInfo.id),
+            mEditController->getParamNormalized(parameterInfo.id)
+         );
+      }
+   }
+
+   return true;
+}
+
+bool VST3Effect::SetAutomationParameters(CommandParameters& parms)
+{
+   using namespace Steinberg;
+
+   if(mComponentHandler == nullptr)
+      return false;
+
+   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   {
+      Vst::ParameterInfo parameterInfo;
+      if(mEditController->getParameterInfo(i, parameterInfo) != kResultOk)
+         continue;
+
+      Vst::ParamValue value;
+      if(parms.Read(std::to_string(parameterInfo.id), &value))
+      {
+         if(mComponentHandler->beginEdit(parameterInfo.id) == kResultOk)
+         {
+            auto cleanup = finally([&]{
+               mComponentHandler->endEdit(parameterInfo.id);
+            });
+            mComponentHandler->performEdit(parameterInfo.id, value);
+         }
+         mEditController->setParamNormalized(parameterInfo.id, value);
+      }
+   }
+
+   return true;
+}
+
+bool VST3Effect::LoadUserPreset(const RegistryPath& name)
+{
+   using namespace Steinberg;
+
+   if(!mEditController)
+      return false;
+
+   if(!PluginSettings::HasConfigValue(*this, PluginSettings::Private, name, wxT("ProcessorState")))
+      return true;
+
+   wxString processorStateStr;
+   if(!GetConfig(*this, PluginSettings::Private, name, wxT("ProcessorState"), processorStateStr, wxEmptyString))
+      return false;
+   auto processorState = owned(PresetsBufferStream::fromString(processorStateStr));
+   if(mEffectComponent->setState(processorState) != kResultOk)
+      return false;
+
+   if(PluginSettings::HasConfigValue(*this, PluginSettings::Private, name, wxT("ProcessorState")))
+   {
+      wxString controllerStateStr;
+      if(!GetConfig(*this, PluginSettings::Private, name, wxT("ControllerState"), controllerStateStr, wxEmptyString))
+         return false;
+      auto controllerState = owned(PresetsBufferStream::fromString(controllerStateStr));
+
+      if(mEditController->setComponentState(processorState) != kResultOk ||
```

Comment:
Is setting component state twice through two different interfaces necessary?  (A eustion to answer from Steinberg documentation)

Reply:
Yes